src/compile/data/bin.ts

```diff
@@ -1,18 +1,32 @@
 import {BinParams, binToString} from '../../bin';
-import {Channel} from '../../channel';
+import {Channel, NONSPATIAL_SCALE_CHANNELS, SPATIAL_SCALE_CHANNELS} from '../../channel';
 import {Config} from '../../config';
 import {field, FieldDef, normalizeBin} from '../../fielddef';
 import {BinTransform} from '../../transform';
-import {Dict, duplicate, extend, flatten, keys, vals} from '../../util';
+import {NOMINAL, ORDINAL} from '../../type';
+import {contains, Dict, duplicate, extend, flatten, keys, vals} from '../../util';
 import {VgBinTransform, VgTransform} from '../../vega.schema';
 import {binFormatExpression} from '../common';
 import {isUnitModel, Model, ModelWithField} from '../model';
 import {DataFlowNode} from './dataflow';
 
-function rangeFormula(model: ModelWithField, fieldDef: FieldDef<string>, channel: Channel, config: Config) {
-    const discreteDomain = model.hasDiscreteDomain(channel);
+/**
+ * Checks whether a fieldDef for a particular channel requires a computed bin range.
+ */
+export function binRequiresRange(fieldDef: FieldDef<string>, channel: Channel) {
+  if (!fieldDef.bin) {
+    console.warn('Only use this method with binned field defs');
+    return false;
+  }
+  if (!contains(NONSPATIAL_SCALE_CHANNELS, channel) && !contains(SPATIAL_SCALE_CHANNELS, channel)) {
```

Comment:
We have `NONSPATIAL_SCALE_INDEX` in channel.ts, maybe you wanna just add type guards and use that as it will be faster and reusable?

Note that currently we also don't have any channel scale without both axis and legend but some of future  nonspatial channel (e.g., fontWeight) _may_ not have legend (simply because it's a pain to implement).

Reply:
Thinking more, in theory, all non-spatial channel can have legends. So I don't really know if we really need this check.

Reply:
Not detail and text, right?

Reply:
Typo:  all non spatial scale channels*

Reply:
So, all non-spatial scale channels.

Reply:
>  all non-spatial channel can have legends
All non-spatial scale channels can have legends

So the following clause
```
!contains(NONSPATIAL_SCALE_CHANNELS, channel) && !contains(SPATIAL_SCALE_CHANNELS, channel)
```
is equivalent to `!isScaleChannel(channel)`

Reply:
For tooltip, this is not really needed, but we might need to modify it bin combine start and `_end`.

Reply:
That's what I'm using now.